src/PlaneWaveBasis.jl

```diff
@@ -194,34 +197,41 @@ build_kpoints(basis::PlaneWaveBasis, kcoords) =
     opIFFT = inv(opFFT)
 
     # Setup kpoint basis sets
+    !variational && @warn(
+        "Non-variational calculations are experimental. " *
+        "Not all features of DFTK may be supported or work as intended."
+    )
     kpoints = build_kpoints(model, fft_size, kcoords, Ecut; variational=variational)
     # kpoints is now possibly twice the size of ksymops. Make things consistent
     if model.n_spin_components == 2
         ksymops = vcat(ksymops, ksymops)
         krange_thisproc = vcat(krange_thisproc, n_kpt .+ krange_thisproc)
+        krange_allprocs = [vcat(range, n_kpt .+ range) for range in krange_allprocs]
     end
 
     # Compute weights
     kweights = [length(symmetries) for symmetries in ksymops]
-    tot_weight = mpi_sum(sum(kweights), mpi_comm)
+    tot_weight = mpi_sum(sum(kweights), comm_kpts)
     kweights = T.(model.n_spin_components .* kweights) ./ tot_weight
-    @assert mpi_sum(sum(kweights), mpi_comm) ≈ model.n_spin_components
+    @assert mpi_sum(sum(kweights), comm_kpts) ≈ model.n_spin_components
 
     # Create dummy terms array for basis to handle
     terms = Vector{Any}(undef, length(model.term_types))
 
     basis = PlaneWaveBasis{T}(
         model, Ecut, kpoints,
-        kweights, ksymops, kgrid, kshift, mpi_comm, krange_thisproc, fft_size,
-        opFFT, ipFFT, opIFFT, ipIFFT,
+        kweights, ksymops, kgrid, kshift, comm_kpts, krange_thisproc, krange_allprocs,
+        fft_size, opFFT, ipFFT, opIFFT, ipIFFT,
         opFFT_unnormalized, ipFFT_unnormalized, opBFFT_unnormalized, ipBFFT_unnormalized,
         terms, symmetries)
     @assert length(kpoints) == length(kweights)
 
     # Instantiate the terms with the basis
-    for (it, t) in enumerate(model.term_types)
-        term_name = string(nameof(typeof(t)))
-        @timing "Instantiation $term_name" basis.terms[it] = t(basis)
+    if instantiate_terms
```

Comment:
Hm that looks like a potential trap. What's the motivation here? To create dummy (non-functional) pwbasis? In which case can't you pass it a dummy model with no terms?

Reply:
Note that in this substantial work is still being done to initialize the basis vectors

Reply:
> can't you pass it a dummy model with no terms?

Not easily. I'd have to create a model from scratch, which I wanted to avoid to reduce coupling between the implementation details of the Model struct and this basis.

 > Note that in this substantial work is still being done to initialize the basis vectors

Yes maybe that optimisation is not worth it, I'm not sure. I wanted to avoid it because if you do checkpointing than this method will be run every SCF step. OTOH probably the communication to gather the k-Points is anyway dominating in this case and the initialisation does not really matter. You generally have a better intuition about this. If you think it's better to avoid this complication and fully initialise I don't mind.

Reply:
I would say let's not optimize now and let's see if it matters afterwards. If you can't afford term creation, you possibly can't afford basis creation anyway

Reply:
ok